### ISSUE TITLE
Error from arangodump

### ISSUE THREAD

**Issue by crs-** [arangodb/arangodb#2203]
I use a database created with version 2 of ArangoDB and imported in version 3.0 of ArangoDB. After upgrading to version 3.1.2 I can not dump the db using arangodump. If I use arangodump for this db like:
`arangodump  --server.username root  --output-directory /backup/arangodump_`date +%Y-%m-%d_%T`/ --server.database MyDB --overwrite true`
it writes an Error:
`2016-12-01T09:31:05Z [2982] ERROR got error from server: HTTP 500 (Internal Server Error): ArangoError 4: internal error (exception location: /var/lib/jenkins/workspace/RELEASE__BuildPackages/arangod/RestHandler/RestReplicationHandler.cpp:3085). Please report this error to arangodb.com`
The document collections and the system collections seem to be dumped corectly but the dump contains no edge collection.
I've tested arangodump with a new created (smaller) databases and therewith it works as expected.
My configuration is:
ArangoDB 3.1.2
OS: Debian 8.6

**Comment by crs-** [comment]
With Version 3.1.3 the following error is written to the log:
`[1516] ERROR caught exception during dump of collection 'MyCollection': Invalid UTF-8 sequence`
All data was inserted by a java tool using the ArangoDB java driver.
The error seems to be caused by one document. The key of this document is known by me, but I can't query for the document with aql.
It will result in the same error message:
`Invalid UTF-8 sequence`

**Comment by crs-** [comment]
After deleting the corrupted document the error message is the same as for Version 3.1.2
`ERROR got error from server: HTTP 500 (Internal Server Error): ArangoError 4: internal error (exception location: /var/lib/jenkins/workspace/RELEASE__BuildPackages/arangod/RestHandler/RestReplicationHandler.cpp:3085). Please report this error to arangodb.com`

**Comment by jsteemann** [comment]
Is there anything else in the server logs that may be related to this import/this error?

**Comment by crs-** [comment]
No that is the only error I see.
But I've tested some AQL queries and I think the error is still a result of the invalid encoding in some documents of the document collection. 
The dump process is interrupted while writing the dump of MyCollection. So no edge collection is written after that. 
I've written a small Java class which changes the encoding of each doument in the collection. I'll post whether it works as expected. 

I think the problem is that the REST-Interface accepts Strings with a wrong encoding that can't be handled by the db-related tools. So I'm able to query for the corrupted documents by using Java/REST but not by using the webinterface.

**Comment by jsteemann** [comment]
I think the JSON and VST REST interfaces do not validate UTF-8 string inputs by default, although they could do this (at the cost of some runtime performance penalty). 
But if you could find out whether everything succeeds with valid UTF-8 data that would be a great help.

**Comment by crs-** [comment]
Ok, I'll try to reconstruct what I've tested within the last days.
1. I tried to query for all documents from MyCollection. It fails with the error message `Invalid UTF-8 string`.
2. I tried to replace alle documents containing a bad encoding with new versions encoded to UTF-8 but several tries fail (after a lot of documents where replaced) with different error messages. The last one was: 
`Response: 400, Error: 1501 - syntax error, unexpected end of query string near '½' at position 1:514 (while parsing) [..]` I couldn't fix that, because the query string does not contain a '½'.
3. I tried to copy all documents with bad encoding into a new collection. This also fails with an error message after a few documents where copied. As I opened the new collection in the webinterface no document was shown but the documents count (lower left corner) told me that the collection was containing some documents.
4. I deleted all documents with a bad encoding and also all edges referencing these documents. Now I can successful query for all documents from MyCollection. But arangodump still fails while dumping MyCollection with the same error message.
5. I've installed the update to Version 3.1.4 and tried to dump the db again. It fails.
But in the info panel of MyCollection I've seen the following:
![infopanel](https://cloud.githubusercontent.com/assets/9804856/21097948/50345f48-c067-11e6-95f7-04cfe94e3de5.jpg)
The number 4491 (under "Deletion" in row "type - Dead") seems to be the number of document with bad encoding I've deleted. 
Is it possible that the documents are still contained in the data files preventing a successful dump?

**Comment by jsteemann** [comment]
Deleting documents from a collection first only writes delete markers into the datafiles. Existing documents are not overwritten in datafiles, so they will remain there until datafiles get compacted.
Compaction can be forced like this:
```js
db.mycollection.truncate();
require("internal").wal.flush(true, true);
try {
  db.mycollection.rotate();
} catch (err) {
}
```
After that, the "dead" count for the collection should go down to zero eventually.

In general it seems that there is an encoding error somewhere in the setup. I am not sure what kind of strings you are passing into ArangoDB or the Java driver, but it seems there is an encoding error somewhere. Can you confirm that all input to ArangoDB is just UTF-8?

**Comment by crs-** [comment]
After doing `db.mycollection.truncate();` the collection is empty.

The collection contains Twitter-Messages.  With the old setup (ArangoDB 2.x, Java driver 2.5.x) I was able to work with the database without any problems. All messages were imported without any cahnges in the encoding. But after I started the import with the new setup (ArangoDB 3.1.x, Java driver 4.1.x) I get the errors described above while dumping the database or querying for messages. The only changes to the import tool were the changes needed by the new driver.

Currently I can not confirm that all input to ArangoDB is just UTF-8.

**Comment by mpv1989** [comment]
Hi,
can you please share some Java code you use to import the documents? I would like to reproduce your problem.

Thanks

**Comment by crs-** [comment]
Ok, here you can find some java code: [DebugTest.zip](https://github.com/arangodb/arangodb/files/651840/DebugTest.zip). (Sorry for the late response.)

The zip-file contains a Java file (needs the ArangoDB Java driver and jackson) and a data file (with bad encoded messages). If you read the data file using the Java code and write it to a document collection you can reproduce the problem. 6 documents will be imported to the collection without any error message but no document is shown if you open the document collection in the web interface and you will not be able to dump the database until you have deleted the imported documents.
The AQL-query:
`FOR message IN MyCollection
RETURN message`
will result in the error message: `Query: Invalid UTF-8 sequence`
If you overwrite the content of the member `message.text = ""` (line 102) all works fine.

**Comment by mpv1989** [comment]
I can't reproduce your problem. With your java code I can write all 6 documents successfully into my db. The web-interface also shows me every document. Besides that your data file looks like you have doubled UTF-8 encoded data.

Maybe you can test the following code. It should insert 256 documents containing a String field created by one byte which is incremented with each document. The documents with id 128 and higher should be written with bad encoding. Check if you see them with the web-interface.

```java
public class EncodingTest {

  private static final String COLLECTION_NAME = "encoding_test";

  @Test
  public void test() {
    final ArangoDB arangoDB = new ArangoDB.Builder().build();
    try {
      arangoDB.db().collection(COLLECTION_NAME).drop();
    } catch (final Exception e) {
    }
    arangoDB.db().createCollection(COLLECTION_NAME);
    byte b = 0x00;
    final Collection<BaseDocument> docs = new ArrayList<>();
    for (int i = 0; i < 256; i++) {
      final BaseDocument doc = new BaseDocument(String.valueOf(i));
      doc.addAttribute("foo", new String(new byte[] { b }));
      b += 1;
      docs.add(doc);
    }
    arangoDB.db().collection(COLLECTION_NAME).insertDocuments(docs);
  }

}
```

**Comment by crs-** [comment]
I've tested your code and it shows the same behavior as my example. After the import no documents are shown in the collection encoding_test. But the counter shows 256 documents.
As next step I've installed a new Debian (version 8.6) on a new server and installed ArangoDB verion 3.1.5 on it. With this new system the behavior is still the same, even after I've changed the system encoding of the Debian-Host to UTF-8.

**Comment by crs-** [comment]
I tried to test your code using ArangoDB 3.1.6 on top of Windows 10 and I got the same results.
Arangodump returns `exception during dump of collection 'encoding_test': Invalid UTF-8 sequence`.

**Comment by mpv1989** [comment]
Hi,
excuse me for not having reported so long. I have successfully reproduce your problem. It was a bug in Java driver version 4.1.0 and was fixed with 4.1.1. I suggest you to switch directly to the latest version 4.1.4 of the java driver, then the problem should no longer occur.

**Comment by crs-** [comment]
Hi,
thanks for your reply. I'm sorry, but this problem really confuses me.
Today I switched from Version 4.1.3 to 4.1.4 (I've done all my previous tests with v. 4.1.3).
I've upgraded Arangodb on my dev system to version 3.1.7 and runned your test code again. Everything seemed to work fine and I was able to see the imported documents in the web interface.
So I started to copy 100 of my corrupt data sets by a small Java program (with driver version 4.1.4) into a new collection. All 100 datasets where shown in the web interface but the query:
`FOR doc IN TestCollection RETURN doc`
still breaks with the error:
`Query: Invalid UTF-8 sequence`
And I'm still not able to dump the collection.

To make sure that no side effects from previous versions disturb my tests, I've installed arangodb 3.1.7 on a new virtual server and started my tests again. Also here I was not able to dump the collection with the 100 documents copied from the old database server. And on both database servers (old and new) I was not able to import of your "encoding_test" documents again so that they are visible in the web interface. I can't say why I was able to see the documensts after one import.

My last try was to delete the documents from the TestCollection one by one. But 5 (visible) documents could not be deleted. After a refresh of my browser window the 5 documents where no longer shown in the web interface, but the counter still says that there are 5 documents in the collection. And the query mentioned above still breaks with the UTF-8 error.

**Comment by mpv1989** [comment]
Hi,
Ok, that is really strange. If I understand you correctly, the 100 documents of yours are shown in the webUI, even on the new virtual server, but could not be queried or dumped? Can you provide these documents to us?

I have written another short test which saves 5 documents with different bad encoded Strings. It would be great if you can run it in your environment and tell us the results (shown in WebUI, query, delete, dump). Thanks!

```java
@Test
public void test2() {
  final ArangoDB arangoDB = new ArangoDB.Builder().build();
  final String collectionName = "bad_encoding";
  try {
    arangoDB.db().collection(collectionName).drop();
  } catch (final Exception e) {
  }
  arangoDB.db().createCollection(collectionName);

  byte[][] data = new byte[][] { 
    new byte[] { (byte) 0x41, (byte) 0x80 },
    new byte[] { (byte) 0x41, (byte) 0xff },
    new byte[] { (byte) 0x42, (byte) 0xff, (byte) 0x70 },
    new byte[] { (byte) 0x43, (byte) 0xff, (byte) 0xff, (byte) 0x07 },
    new byte[] { (byte) 0x44, (byte) 0xff, (byte) 0xff, (byte) 0xff, (byte) 0x07 }
  };

  Collection<BaseDocument> docs = new ArrayList<>();
  for (int i = 0; i < data.length; i++) {
    final BaseDocument doc = new BaseDocument(String.valueOf(i));
    doc.addAttribute("foo", new String(data[i]));
    docs.add(doc);
  }
  arangoDB.db().collection(collectionName).insertDocuments(docs);
}
```

**Comment by crs-** [comment]
Hi,
yes the documents are shown sometimes in the webUI. I can't say in wich case. In most cases when I switch to a page with a bad document, the webUI will not display the document. I have veryfied this by searching for the key of a bad document.
If I query for a document with a bad character I get the UTF-8 error. The same happens if I dump the collection with "bad documents". But I can query for all other attributes (without bad chars) of the bad document.

Today I've done some new tests independent from my original database.
I'll try to declare it in detail and I hope you can reproduce the problem. I have:
1. created a new LXC-Container (debian 8.6 x64)
2. installed Arangodb 3.1.8 by apt-get according to the instructions on the Arangodb download page
3. configured /etc/arangodb3/arangod for endpoint http://0.0.0.0:8529 and restarted the arangod
4. created a database "Debug" with owner "root"

On my client pc (Win 10 x64 with Eclipse + Java1.8.0_111 x64 - I've compiled both tests with  32-bit version of Java1.8.0_111 too. It makes no difference. -) I:
1. created a new Java project that uses Maven (no changes in the charset encoding was made)
2. added arangodb-java-driver: 4.1.4 as dependency to pom.xml
3. created a class with the code you can find here: [ImportTest.zip](https://github.com/arangodb/arangodb/files/699243/ImportTest.zip) 

This code creates 54 collections in the database "Debug". 27 of them (Name: `Test_single_[index]`) contain only one document with one attribute "foo" and one string from the values-array.
The other 27 collections (Name: `Test_all_[index]`) contain documents for all values from the values-array with an index <= [index]. 
All collections with [index] <= 20 look good. But the collections with [index] > 20 (starting with character "§") show a strange behaviour. Sometimes documents form another collection (opened before) will be displayed in the webUI as content of a "bad colletion" but if the browser window (tested with Firefox and Opera) will be refreshed the "No documents"-String will be displayed instead of the documents contained in the collection. Also the "good documents" will not be displayed anymore for this collection. But the couter (bottom - left) shows the right count of documents for the collection.  I've changed the order of the values-array several times and AFAICT the chars `"§", "Ö", "Ä", "Ü", "ä", "ö", "ü"` seems to be problematic.
I hope you can reproduce my tests. I'll do my best to answer to possible questions.

Summarized I get the following results:
For my tests:
- webUI: the content of collections with bad documents [index] >20 is not displayed correctly
- dump: breaks on document collection 'Test_all_21' (first collection that contains a bad encoding) [(whole output)](https://github.com/arangodb/arangodb/files/699290/Output_ImportTest_Dump.zip)
- query (`FOR doc IN Test_[all || single]_[index] RETURN doc`): Works for alls collections with [index] <=20 and results in the UTF-8 error for the other collections 
- delete (`FOR doc IN Test_all_27 REMOVE doc IN Test_all_27`): successful (no dead documents)
- manual delete (by switching between a collection which content will be displayed ("Test_all_20") an one which content will not be displayed ("Test_all_21")): I could see the documents from Test_all_20 while displaying the content of Test_all_21 and I was able to delete some documents but not all.

For your code:
- webUI: the content of the collection "bad_encoding" is not visible: ![screenshot](https://cloud.githubusercontent.com/assets/9804856/21854996/2196b686-d81d-11e6-9a0c-35159abd0c74.jpg)
- dump: breaks with UTF-8 error [(whole output)](https://github.com/arangodb/arangodb/files/699304/Output_EncTest_Dump.zip)
- query (`FOR doc IN bad_encoding RETURN doc`): results in the UTF-8 error 
- delete only with query: (`FOR doc IN bad_encoding REMOVE doc IN bad_encoding`): successful (no dead documents) 

While trying to manual delete some documents from collection Test_all_21 I accidentally opened one of the documents that couldn't be deleted and it looked to me like an edge document (But I've done all test exclusively with document collections):
![screenshot2](https://cloud.githubusercontent.com/assets/9804856/21856529/9e1e18b6-d822-11e6-83ce-43ab55faa924.jpg)

**Comment by mpv1989** [comment]
Thanks for the source-code.

Today I released version 4.1.5 of the java-driver including a bugfix which should be solve your problems. It should be available within the next hours on maven-central repository.

You can already try it with the snapshot version (4.1.5-SNAPSHOT). To use snapshots add the staging repository of oss.sonatype.org to your pom.xml:
```xml
<repositories>
  <repository>
    <id>arangodb-snapshots</id>
    <url>https://oss.sonatype.org/content/groups/staging</url>
  </repository>
</repositories>
```

**Comment by crs-** [comment]
Works great, thanks!
I was able to copy the documents with the bad encoding to a new collection and now I can query for the documents with AQL without getting an error.
